README.md
Payline extension for wooCommerce
====================================

This plugin allows the merchant to connect his store to Payline payment gateway.
Payline offers many payment means (Visa, Mastercard, American Express, Paypal, JCB,...) and a powerfull anti-fraud system.

Docs
====

All information available on https://docs.payline.com/display/DT/Plugin+WooCommerce


About
=====

Requirements
------------

This extension requires at least wooCommerce 3.0
It's tested up to WooCommerce 4.9.2.


Author
------

* Fabien SUAREZ - (https://www.monext.fr/retail)
* Vincent Pietri (https://www.tbdgroup.com/)
* Nicolas MOLLET (https://github.com/thermesmarins)
* Timothe BORDIGA (https://github.com/roux1max)


License
-------

Payline is licensed under the LGPL-3.0+ License


CHANGELOG.md
Changelog
-------

---
* 1.4 - 2021/01/29
  * Add widget integration
  * Add NX and REC payment methods
  * Add Refund online feature
  * Improve user experience
  * Compliance with WooCommerce 4.9.2
  * Compliance with Payline PHP SDK 4.59
  
---
* 1.3.7 - 2020/12/01
  * Feature - WooCommerce 3.x compatibility (not compatible anymore with WooCommerce versions below 2.6)
  * Transaction id compatibility, Translation files
  * Fix on token get data (versus paylinetoken)
---
* 1.3.6 - 2018/01/02  
  * Fix - Truncate buyer data before send it to Payline.
---
* 1.3.5 - 2017/04/04
  * Feature - send buyer info mandatoty for Cetelem 3x / 4x
---     
* 1.3.4 - 2017/02/27  
  * Fix - languages files
---
* 1.3.3 - 2016/08/26  
  * Feature - order/token association. Prevents conflicts between payment sessions.
---
* 1.3.2 - 2016/08/04  
  * Fix - Truncate order details product name to 50 characters before send it to Payline.
---
* 1.3.1 - 2015/12/09  
  * Feature - compliance with Payline PHP library v4.43
---
* 1.3 - 2015/02/27  
  * Feature - compliance with wc 2.3 and over


class-wc-gateway-payline.php
<?php


/**
 * Payline module for WooCommerce
 *
 * @class 		WC_Payline
 * @package		WooCommerce
 * @category	Payment Gateways
 *
 * WC tested up to: 4.0.1
 */


class WC_Gateway_Payline extends WC_Abstract_Payline {

    protected $paymentMode = 'CPT';

    public $id = 'payline';

    public $method_title = 'Payline CPT';


    /**
     * @param WC_Refund|bool|WC_Order $order
     * @return mixed|void
     */
    protected function getWebPaymentRequest(WC_Order $order) {

        $requestParams = parent::getWebPaymentRequest($order);

        do_action('payline_before_do_web_payment', $requestParams, $this);

        return $requestParams;
    }

    /**
     * @param WC_Order $order
     * @param array $res
     * @return false
     */
    protected function paylineCancelWebPaymentDetails(WC_Order $order, array $res) {
        return false;
    }

    /**
     * @param WC_Order $order
     * @param array $res
     * @return bool
     */
    protected function paylineSuccessWebPaymentDetails(WC_Order $order, array $res) {

        if($res['result']['code'] == '00000') {
            $orderId = $order->get_id();

            // Store transaction details
            update_post_meta((int) $orderId, 'Transaction ID', $res['transaction']['id']);
            update_post_meta((int) $orderId, 'Card number', $res['card']['number']);
            update_post_meta((int) $orderId, 'Payment mean', $res['card']['type']);
            update_post_meta((int) $orderId, 'Card expiry', $res['card']['expirationDate']);

            update_post_meta((int) $orderId, '_contract_number', $res['payment']['contractNumber']);
            $order->payment_complete($res['transaction']['id']);
            $order->update_status('completed', 'Payment validated');
            return true;
        }
        return false;
    }






}


class-wc-gateway-payline-nx.php
<?php

use Payline\PaylineSDK;

/**
 * Payline module for WooCommerce
 *
 * @class 		WC_Payline
 * @package		WooCommerce
 * @category	Payment Gateways
 *
 * WC tested up to: 4.0.1
 */


class WC_Gateway_Payline_NX extends WC_Abstract_Recurring_Payline_NX {


    protected $paymentMode = 'NX';

    public $id = 'payline_nx';

    public $method_title = 'Payline N Fois';

    function init_form_fields()
    {
        parent::init_form_fields();

        $this->form_fields['billing_left'] = array(
            'title' => __('Billing left', 'payline'),
            'default' => '3',
            'type' => 'text',
            'description' => __('Recurring billing number. Mandatory field', 'payline')
        );
    }

    /**
     * @see https://docs.payline.com/pages/viewpage.action?pageId=747147142
     * @param WC_Refund|bool|WC_Order $order
     * @return mixed|void
     */
    protected function getWebPaymentRequest(WC_Order $order) {

        $requestParams = parent::getWebPaymentRequest($order);

        $billingLeft = (int)$this->settings['billing_left'];
        if(empty($billingLeft) or $billingLeft<2) {
            $billingLeft = 2;
        }
        $requestParams['recurring']['billingLeft'] = $billingLeft; // Nombre d’échéance

        $totalAmount = $requestParams['payment']['amount'];
        $recurringAmount = round(( $totalAmount / 100 ) / $billingLeft) * 100;

        $requestParams['recurring']['firstAmount'] = round($totalAmount - ($recurringAmount * ($billingLeft-1)));


        $requestParams['recurring']['amount'] = $recurringAmount;
        $requestParams['recurring']['billingCycle'] = $this->settings['billing_cycle'];


        $requestParams['recurring']['billingDay'] = '01'; //  [01 à 30]

        //$today    = current_time( 'd/m/Y' );
        $requestParams['recurring']['startDate'] = ''; // dd/mm/yyyy

        do_action('payline_before_do_web_payment_nx', $requestParams, $this);

        return $requestParams;
    }

    /**
     * Check if the gateway is available for use.
     *
     * @return bool
     */
    public function is_available() {

        $is_available = parent::is_available();
        $cart = WC()->cart;
        if ($is_available && $cart) {
            return $this->settings['billing_left'] > 0;
        }
        return $is_available;
    }

}


class-wc-gateway-payline-rec.php
<?php

use Payline\PaylineSDK;

/**
 * Payline module for WooCommerce
 *
 * @class 		WC_Payline
 * @package		WooCommerce
 * @category	Payment Gateways
 *
 * WC tested up to: 4.0.1
 */


class WC_Gateway_Payline_REC extends WC_Abstract_Recurring_Payline_NX {


    protected $paymentMode = 'REC';

    public $id = 'payline_rec';

    public $method_title = 'Payline par Abonnement';


    /**
     * Check if the gateway is available for use.
     *
     * @return bool
     */
    public function is_available() {

        $is_available = parent::is_available();
        $cart = WC()->cart;
        if ($is_available && $cart) {
            $eligibleIds = !empty($this->settings['eligible_product_ids']) ? explode(";", $this->settings['eligible_product_ids']) : array();
            foreach ( $cart->get_cart() as $cart_item_key => $cart_item ) {
                $product = $cart_item['data'];
                if(!in_array($product->get_id(), $eligibleIds)) {
                    return false;
                }
            }
        }
        return $is_available;
    }

    /**
     *
     */
    function init_form_fields()
    {
        parent::init_form_fields();

        $this->form_fields['max_records'] = array(
            'title' => __('Maximum records', 'payline'),
            'default' => '12',
            'type' => 'text',
            'description' => __('Set a number maximum of records or leave empty', 'payline')
        );

        $this->form_fields['eligible_product_ids'] = array(
            'title' => __('Eligible product ids', 'payline'),
            'default' => '',
            'type' => 'text',
            'description' => __('Define a list of product ids that can be payed with REC. Values must be separated by ;', 'payline')
        );
    }

    /**
     * @see https://docs.payline.com/pages/viewpage.action?pageId=747147142
     * @param WC_Refund|bool|WC_Order $order
     * @return mixed|void
     */
    protected function getWebPaymentRequest(WC_Order $order) {

        $requestParams = parent::getWebPaymentRequest($order);

        $totalAmount = $requestParams['payment']['amount'];
        $requestParams['recurring']['firstAmount'] = $totalAmount;

        $requestParams['recurring']['amount'] = $totalAmount;
        $requestParams['recurring']['billingCycle'] = $this->settings['billing_cycle'];


        $requestParams['recurring']['billingDay'] = '01'; //  [01 à 30]

        //$today    = current_time( 'd/m/Y' );
        $requestParams['recurring']['startDate'] = ''; // dd/mm/yyyy


        if($this->settings['billing_cycle'] && $this->settings['max_records']) {
            $numberDaysToLastDate = $this->getDaysForCycles($this->settings['billing_cycle']) * $this->settings['max_records'];
            $requestParams['recurring']['endDate'] = current_datetime()->modify( '+' .$numberDaysToLastDate. ' day' )->format( 'd/m/Y' );;

        }

        do_action('payline_before_do_web_payment_rec', $requestParams, $this);

        return $requestParams;
    }
}


class-wc-abstract-recurring-payline.php
<?php

use Payline\PaylineSDK;

/**
 * Payline module for WooCommerce
 *
 * @class 		WC_Payline
 * @package		WooCommerce
 * @category	Payment Gateways
 *
 * WC tested up to: 4.0.1
 */

abstract class WC_Abstract_Recurring_Payline_NX extends WC_Abstract_Payline {


    protected $paymentMode = 'NX';

    public $id = 'payline_nx';

    public $method_title = 'Payline N Fois';

    function init_form_fields() {
        parent::init_form_fields();

        $this->form_fields['payment_action']['default'] = '101';
        $this->form_fields['payment_action']['options'] = array(
            '101' => __('Authorization + Capture', 'payline')
        );

        /*
         * Payment Settings
         */
        $this->form_fields[$this->id] = array(
            'title' => strtoupper(__($this->method_title, 'payline' )),
            'type' => 'title'
        );

        $this->form_fields['billing_cycle'] = array(
            'title' => __('Payment frequencies', 'payline'),
            'type' => 'select',
            'default' => '40',
            'options' => array(
                '10' => __('One transaction per day', 'payline'),
                '20' => __('One transaction every seven days', 'payline'),
                '30' => __('Two transactions per month', 'payline'),
                '40' => __('One transaction per month', 'payline'),
                '50' => __('One transaction every two months', 'payline'),
                '60' => __('One transaction every three months', 'payline'),
                '70' => __('One transaction every six months', 'payline'),
                '80' => __('One transaction per year', 'payline'),
                '90' => __('A transaction every two years', 'payline')
            )
        );

    }

    protected function getDaysForCycles($code)
    {
        $cyclesDays = array('10' => 1,
            '20' => 7,
            '30' => 15,
            '40' => 30,
            '50' => 60,
            '60' => 90,
            '70' => 180,
            '80' => 360,
            '90' => 720
        );

        return !empty($cyclesDays[$code]) ? $cyclesDays[$code] : 0;
    }


    /**
     * @param WC_Order $order
     * @param array $res
     * @return false
     */
    protected function paylineCancelWebPaymentDetails(WC_Order $order, array $res) {
        return false;
    }

    /**
     * @param WC_Order $order
     * @param array $res
     * @return bool
     */
    protected function paylineSuccessWebPaymentDetails(WC_Order $order, array $res) {

        if($res['result']['code'] == '02500') {
            $orderId = $order->get_id();

            // Store transaction details
            update_post_meta((int) $orderId, 'Transaction ID', $res['transaction']['id']);

            update_post_meta((int) $orderId, '_contract_number', $res['payment']['contractNumber']);

            $order->payment_complete($res['transaction']['id']);

            $order->update_status('completed', 'First payment validated');
            return true;
        }

        return false;
    }

}


woocommerce-payline.php
<?php
/**
 * Plugin Name: Payline
 * Plugin URI: https://docs.payline.com/display/DT/Plugin+WooCommerce
 * Description: integrations of Payline payment solution in your WooCommerce store
 * Version: 1.4
 * Author: Monext
 * Author URI: http://www.monext.fr
 * License: LGPL-3.0+
 * GitHub Plugin URI: https://github.com/PaylineByMonext/payline-woocommerce/
 * Github Branch: master
 * WC tested up to: 4.9.2
 * 
 *  Copyright 2017  Monext  (email : support@payline.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 2, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

if (!defined('ABSPATH')) exit;

define('WCPAYLINE_PLUGIN_URL', plugin_dir_url(__FILE__));

function woocommerce_payline_activation() {
	if (!is_plugin_active('woocommerce/woocommerce.php')) {
		deactivate_plugins(plugin_basename(__FILE__));

		load_plugin_textdomain('payline', false, dirname(plugin_basename(__FILE__)) . '/languages/');
		
		$message = sprintf(__('Sorry! In order to use WooCommerce %s Payment plugin, you need to install and activate the WooCommerce plugin.', 'payline'), 'Payline');
		wp_die($message, 'WooCommerce Payline Gateway Plugin', array('back_link' => true));
	}
}
register_activation_hook(__FILE__, 'woocommerce_payline_activation');


/**
 * inserts class gateway
 */
function woocommerce_payline_init() {
	// Load translation files
	load_plugin_textdomain('payline', false, dirname(plugin_basename(__FILE__)) . '/languages/');

    if ( ! class_exists( 'WC_Abstract_Payline', false ) ) {
        include_once 'class-wc-abstract-payline.php';
    }

    if ( ! class_exists( 'WC_Abstract_Recurring_Payline_NX', false ) ) {
        include_once 'class-wc-abstract-recurring-payline.php';
    }
	
	if (!class_exists('WC_Gateway_Payline')) {
		require_once 'class-wc-gateway-payline.php';
	}

    if (!class_exists('WC_Gateway_Payline_NX')) {
        require_once 'class-wc-gateway-payline-nx.php';
    }

    if (!class_exists('WC_Gateway_Payline_REC')) {
        require_once 'class-wc-gateway-payline-rec.php';
    }

	require_once 'vendor/autoload.php';
}
add_action('woocommerce_init', 'woocommerce_payline_init');



/**
 * adds method to woocommerce methods
 * @param $methods
 * @return mixed
 */
function woocommerce_payline_add_method($methods) {
    $methods[] = 'WC_Gateway_Payline';
    $methods[] = 'WC_Gateway_Payline_NX';
    $methods[] = 'WC_Gateway_Payline_REC';

    return $methods;
}
add_filter('woocommerce_payment_gateways', 'woocommerce_payline_add_method');


/**
 * add a link from plugin list to parameters
 * @param $links
 * @param $file
 * @return mixed
 */
function woocommerce_payline_add_link($links, $file) {
	$links[] = '<a href="'.admin_url('admin.php?page=wc-settings&tab=checkout&section=payline').'">' . __('Settings CPT') .'</a><br />';
    $links[] = '<a href="'.admin_url('admin.php?page=wc-settings&tab=checkout&section=payline_nx').'">' . __('Settings NX') .'</a>';
    $links[] = '<a href="'.admin_url('admin.php?page=wc-settings&tab=checkout&section=payline_rec').'">' . __('Settings REC') .'</a>';
	return $links;
}
add_filter('plugin_action_links_'.plugin_basename(__FILE__), 'woocommerce_payline_add_link',  10, 2);

/**
 * @param $statuses
 * @return mixed
 */
function woocommerce_payline_add_custom_status_for_order_again( $statuses ){

    if(isset( $_GET['order_again'], $_GET['_wpnonce'], $_GET['payline_cancel']) && wp_verify_nonce( wp_unslash( $_GET['_wpnonce'] ), 'woocommerce-order_again' ) ) {
        $statuses[] = 'cancelled';
        $statuses[] = 'failed';
        $statuses[] = 'pending';

    }

    return $statuses;
}
add_filter( 'woocommerce_valid_order_statuses_for_order_again', 'woocommerce_payline_add_custom_status_for_order_again', 10, 1 );


/**
 * @param $available_gateways
 * @return mixed
 */
function woocommerce_payline_enable_gateway_order_pay( $available_gateways ) {
    if ( is_checkout() && is_wc_endpoint_url( 'order-pay' ) ) {
        unset( $available_gateways['payline'] );
        unset( $available_gateways['payline_nx'] );
        unset( $available_gateways['payline_rec'] );
    }

    return $available_gateways;
}

add_filter( 'woocommerce_available_payment_gateways', 'woocommerce_payline_enable_gateway_order_pay' );